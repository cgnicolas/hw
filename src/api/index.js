const url = 'http://localhost:3001';

const API = {
  call(url, method, body = {}) {
    const data = {
      method,
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    };
    if (Object.keys(body).length > 0) {
      data.body = JSON.stringify(body);
    }
    return fetch(url, data).then((response) => {
      return response.json();
    });
  },

  get() {
    const path = '/getComments';
    return this.call(url + path, 'get');
  },

  post(body = {}) {
    const path = '/createComment';
    return this.call(url + path, 'post', body);
  },

  delete() {
    const path = '/deleteComments';
    return this.call(url + path, 'delete');
  },
};

export default API;
