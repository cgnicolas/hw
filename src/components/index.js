export { default as Form } from './Form';
export { default as CommentsList } from './CommentsList';
