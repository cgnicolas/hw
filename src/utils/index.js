export const sortComments = (comments) => {
    return comments.sort((a, b) => {
        const { created: aTimestamp } = a;
        const { created: bTimestamp } = b;
        const aDate = new Date(aTimestamp);
        const bDate = new Date(bTimestamp);
        return bDate.getTime() - aDate.getTime();
      });
}