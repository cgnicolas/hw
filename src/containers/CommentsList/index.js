import React from 'react';
import { CommentsList } from '../../components';

export default function CommentsListContainer({ comments = [], onRefresh, loading, error}) {

  const displayComments = () => {
      if(error) {
        return (
          <p>Something went wrong loading the comments. Please try again.</p>
        )
      }
      if(!!comments.length || loading){
        return (
          <React.Fragment>
            {loading && <CommentsList.Skeleton/>}
            {comments.map((comment) => {
              return <CommentsList.Comment {...comment} key={comment.id} />;
            })}
          </React.Fragment>
        )
      }
      return (
        <p>No posts to display. You should change that!</p>
      )
  }

  return (
    <CommentsList onRefresh={onRefresh}>
      {displayComments()}
    </CommentsList>
  );
}
