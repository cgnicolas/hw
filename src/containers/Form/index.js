import React, { useState } from 'react';
import { Form } from '../../components';
import { v4 as uuidv4 } from 'uuid';
export default function FormContainer({ onSubmit, loading }) {
  const [name, setName] = useState('');
  const [message, setMessage] = useState('');
  const formSubmission = (e) => {
    e.preventDefault();
    onSubmit({
      message,
      name,
      id: uuidv4(),
    }).then(() => {
      setMessage('');
    });
  };

  return (
    <Form onSubmit={formSubmission}>
      <Form.Input
        name="Name"
        placeholder="Who are you posting as?"
        value={name}
        id="user-name"
        onChange={({ target: { value } }) => setName(value)}
      />
      <Form.MessageContainer>
        <Form.MessageInput
          placeholder="Join the discussion!"
          value={message}
          id="user-message"
          onChange={({ target: { value } }) => setMessage(value)}
        />
        <Form.Button disabled={loading || !message}>Post</Form.Button>
      </Form.MessageContainer>
    </Form>
  );
}
