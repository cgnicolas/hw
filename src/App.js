import React, { useEffect, useState } from 'react';
import { CommentsList, Form } from './containers/';
import API from './api';
import { sortComments } from './utils';

function App() {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [comments, setComments] = useState([]);

  /*
    Fetch Existing Comments
  */
  const getAndSetComments = () => {
    setLoading(true);
    API.get().then((response) => {
      const sortedComments = sortComments(response);
      setComments(sortedComments);
    })
    .catch((err) => {
      setError(err);
    })
    .finally(() => {
      setLoading(false);
    });
  }

  useEffect(() => {
    getAndSetComments();
  }, []);

  const handleCreate = (commentBody) => {
    setLoading(true);
    return API.post(commentBody)
      .then(() => {
        return getAndSetComments();
      })
      .catch((err) => {
        setError(err);
        return err;
      })
  };

  const refreshComments = () => {
    getAndSetComments();
  }

  return (
    <div className="bg-blue-100 h-screen w-screen">
      <div className="m-auto flex flex-col w-1/2 h-full min-w-xs md:px-0">
        <Form loading={loading} onSubmit={handleCreate} error={error} />
        <CommentsList comments={comments} loading={loading} error={error} onRefresh={refreshComments} />
      </div>
    </div>
  );
}

export default App;
