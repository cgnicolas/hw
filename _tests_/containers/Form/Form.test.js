import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from '@testing-library/react';
import { Form } from '../../../src/containers';

const mockSetState = jest.fn();
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn((init) => [init, mockSetState]),
}));

describe('<Form/>', () => {
  it('should render successfully', () => {
    const renderedModule = shallow(<Form />);
    expect(renderedModule).toMatchSnapshot();
  });

  describe('set by onChange', () => {
    const event = {
      target: { value: 'Wow such a good mocked value! Impressed huh?' },
    };
    describe('setName', () => {
      it('should set the name of the user', () => {
        const renderedModule = mount(<Form />);
        console.log(renderedModule.find('input#user-name').debug());
        act(() =>
          renderedModule.find('input#user-name').props().onChange(event)
        );
        expect(mockSetState).toHaveBeenCalledWith(event.target.value);
      });
    });

    describe('setMessage', () => {
      it('should set the message input by the user', () => {
        const renderedModule = mount(<Form />);
        console.log(renderedModule.find('textarea#user-message').debug());
        act(() =>
          renderedModule.find('textarea#user-message').props().onChange(event)
        );
        expect(mockSetState).toHaveBeenCalledWith(event.target.value);
      });
    });
  });

  describe('formSubmission', () => {
    const props = {
      onSubmit: jest.fn(() => Promise.resolve()),
    };
    it('should submit the form', () => {
      const renderedModule = shallow(<Form {...props} />);
      renderedModule
        .find('FormComponent')
        .props()
        .onSubmit({ preventDefault: jest.fn() });
    });
  });
});
