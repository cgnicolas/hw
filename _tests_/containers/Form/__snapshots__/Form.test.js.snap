// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form/> should render successfully 1`] = `
<FormComponent
  onSubmit={[Function]}
>
  <Component
    id="user-name"
    name="Name"
    onChange={[Function]}
    placeholder="Who are you posting as?"
    value=""
  />
  <Component>
    <Component
      id="user-message"
      onChange={[Function]}
      placeholder="Join the discussion!"
      value=""
    />
    <Component
      disabled={true}
    >
      Post
    </Component>
  </Component>
</FormComponent>
`;
